qiskit_experiments/library/randomized_benchmarking/sampling_utils.py

```diff
@@ -0,0 +1,359 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Utilities for sampling layers in randomized benchmarking experiments
+"""
+
+import warnings
+import math
+from abc import ABC, abstractmethod
+from typing import Optional, Union, List, Tuple, Sequence, TypeVar, NamedTuple
+from collections import defaultdict
+
+from numpy.random import Generator, default_rng, BitGenerator, SeedSequence
+import numpy as np
+
+from qiskit.circuit import Instruction
+from qiskit.exceptions import QiskitError
+
+from .clifford_utils import CliffordUtils, _CLIFF_SINGLE_GATE_MAP_1Q
+
+GateTypeT = TypeVar("GateTypeT", str, int, Instruction)
+
+
+class GateInstruction(NamedTuple):
+    """Named tuple class for sampler output."""
+
+    # the list of qubits to apply the operation on
+    qargs: tuple
+    # the operation to apply
+    op: GateTypeT
+
+
+class BaseSampler(ABC):
+    """Base class for samplers that generate circuit layers based on a defined
+    algorithm and gate set. Subclasses must implement the ``__call__()`` method
+    which outputs a number of circuit layers."""
+
+    def __init__(
+        self,
+        gate_distribution=None,
+        coupling_map: Optional[List[List[int]]] = None,
+        seed: Optional[Union[int, SeedSequence, BitGenerator, Generator]] = None,
+    ):
+        """Initializes the sampler.
+
+        Args:
+            seed: Seed for random generation.
+            gate_distribution: The gate distribution for sampling.
+        """
+        self._gate_distribution = gate_distribution
+        self._coupling_map = coupling_map
+        self.seed = default_rng(seed)
+
+    @property
+    def coupling_map(self):
+        """The coupling map of the system to sample over."""
+        return self._coupling_map
+
+    @coupling_map.setter
+    def coupling_map(self, coupling_map):
+        self._coupling_map = coupling_map
```

Comment:
This is used in `MirrorRB._set_distribution_options` to set the coupling map because we allow users to update the backend until `circuits()` is called.